Redesign comparison structure of TCK implementation

Mats-SX opened the pull request:
This PR introduces the creation of matchers instead of actual objects. These matchers are created from parsing the TCK features, and contains logic that will determine whether a real object returned from the database fulfils the expectations of the TCK feature.

Based on #6470, and should be merged only after that.

Mats-SX committed:
Update grammar to support proper path descriptions

Mats-SX committed:
Rework parsed representation of feature results

Instead of constructing implementations of Java API interfaces (Node,
Relationship, Path), the parser listener will create matcher objects, which
will contain logic that determines whether they are equal to objects returned
from withing Neo4j.

This commit plugs in the matcher-creator, and adds matcher classes for
integers and floats.

Mats-SX committed:
Use matchers for remaining primitives

Mats-SX committed:
Use matchers for containers

Mats-SX committed:
Use matcher for nodes

Mats-SX committed:
Add unit tests for all matchers

Make NullMatcher a static field on ValueMatcher.

Mats-SX committed:
Add unit test for `NodeMatcher`

Restructure `*MatcherTests` to share code more easily.

Mats-SX committed:
Add matcher for relationships

Mats-SX committed:
Remove obsolete code

Mats-SX committed:
Implement matchers for the upper-level structures

Mats-SX committed:
Add parsing of Paths

Mats-SX committed:
Stop closing results on last call of `next()`

Generally, a user may want to act on the last thing returned from the result,
which might require the transaction to be open. This enables that behaviour,
now requiring another call to `hasNext()` for closing to happen.

Mats-SX committed:
Make Cypher return 64-bit integers uniformly

Mats-SX committed:
Ensure `override` on all overridden functions

Mats-SX committed:
Plug in the matcher-based comparison

- Now also solves expected results containing paths
- Handle sorted/unsorted results
- Fix a few bugs in `.feature` files

Mats-SX committed:
Improve description of empty result expectations

Mats-SX committed:
Remove unused code

Mats-SX committed:
Make unit tests run also through maven

Mats-SX commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/ListMatcher.java:
```diff
@@ -0,0 +1,59 @@
+                {
+                    ValueMatcher valueMatcher = list.get( i );
+                    Object value1 = realList.get( i );
+                    match &= valueMatcher.matches( value1 );
```
Could inline this.

Mats-SX commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/Matcher.java:
```diff
@@ -0,0 +1,30 @@
+{
+    boolean matches( T value );
+
+    interface OrderedMatcher<T> extends Matcher<T>
```
Not sure if these two interfaces are necessary. They might add a bit of clarity as for the purpose of the three implementing classes, but I'm not certain of the value.

thobe replied:
`OrderedMatcher` should not extend `Matcher`, the inheritance does not appear to serve any purpose.

Mats-SX commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/NodeMatcher.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public String toString()
+    {
+        return "NodeMatcher for a node with labelNames: " + labelNames + " and properties: " + propertyMatcher;
```
This is too lengthy to be helpful.

Mats-SX commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/PathMatcher.java:
```diff
@@ -0,0 +1,85 @@
+                {
+                    Iterator<Relationship> relationships = path.relationships().iterator();
+                    for (PathLinkMatcher pathLink : pathLinks) {
+                        pathLink.matches( relationships.next() );
```
This tests both end nodes of all relationships in the path. Is that enough to say that the path matches?

Mats-SX commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/PathMatcher.java:
```diff
@@ -0,0 +1,85 @@
+    {
+        if (pathLinks.isEmpty())
+        {
+        return "PathMatcher for " + singleNodePath;
```
Bad code style.

Mats-SX commented on community/cypher/compatibility-suite/src/test/java/cypher/FeatureSuiteTest.java:
```diff
@@ -49,14 +49,15 @@
+//    } )
+//    public static class CostCompiled
+//    {
+//    }
```
We should probably just remove this code. It can be retrieved from the history whenever we start working on the runtime again.

Mats-SX committed:
Improve `toString()` to be more helpful

- Improve code style

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/ListMatcher.java:
```diff
@@ -0,0 +1,57 @@
+            {
+                for ( int i = 0; i < list.size(); ++i )
+                {
+                    match &= list.get( i ).matches( realList.get( i ) );
```
just fail fast and skip the `matches` flag

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/MapMatcher.java:
```diff
@@ -0,0 +1,64 @@
+            {
+                for ( String key : expectedKeys )
+                {
+                    match &= map.get( key ).matches( realMap.get( key ) );
```
just fail fast and skip the `matches` flag

thobe replied:
just fail fast and skip the `match` flag

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/MapMatcher.java:
```diff
@@ -0,0 +1,64 @@
+
+            if ( match )
+            {
+                for ( String key : expectedKeys )
```
why not iterate over the entries of `this.map` instead, to avoid one map.get() call per iteration

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/MapMatcher.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public String toString()
+    {
+        return "MapMatcher {" + map.toString();
```
remove the space and the `{`

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/NodeMatcher.java:
```diff
@@ -0,0 +1,66 @@
+            int nbrOfLabels = 0;
+            for ( Label l : labels )
+            {
+                match &= labelNames.contains( l.name() );
```
just fail fast instead of using the `match` flag (unless we want to _report_ all things that didn't match)

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/PathLinkMatcher.java:
```diff
@@ -0,0 +1,73 @@
+        return "PathLinkMatcher "
+               + leftNode + (outgoing ? "-" : "<-")
+               + relMatcher + (outgoing ? "->" : "-")
+               + rightNode;
```
it would have been nice with some sort of bracket that grouped the representation

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/PathMatcher.java:
```diff
@@ -0,0 +1,88 @@
+            {
+                boolean matches = path.length() == 0;
+                matches &= singleNodePath.matches( path.startNode() );
+                matches &= singleNodePath.matches( path.endNode() );
```
if the length is `0`, `path.startNode()` and `path.endNode()` is the same node, do we really need to do a full matching? could we not do some other thing to assert that invariant?

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/PathMatcher.java:
```diff
@@ -0,0 +1,88 @@
+                    {
+                        pathLink.matches( relationships.next() );
+                    }
+                }
```
matching should fail if `pathLink` doesn't match...

thobe commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/PathMatcher.java:
```diff
@@ -0,0 +1,88 @@
+                }
+                return matches;
+            }
+        }
```
It should be possible to just have one case here, in the case of `path.length() == 0` the iterating case should "degrade" into the other case.

Mats-SX commented on community/cypher/compatibility-suite/src/main/java/cypher/feature/parser/matchers/PathMatcher.java:
```diff
@@ -0,0 +1,88 @@
+                    Iterator<Relationship> relationships = path.relationships().iterator();
+                    for ( PathLinkMatcher pathLink : pathLinks )
+                    {
+                        pathLink.matches( relationships.next() );
```
Check the return value.

Mats-SX committed:
Use fail-fast pattern in matchers

- Fixes bug where paths did not consider whether pathlinks matched
- Fixes bug where paths were parsed in reverse order
